### PR TITLE
Coordinated Tandem Ramp: Sigma Decay Synced With Loss Ramp

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current --tandem_ramp ramps the tandem OOD loss weight from 0→1 over training. The --aug_gap_stagger_sigma=0.02 is applied constantly. These two training dynamics are uncoordinated: the model receives constant geometric diversity from epoch 1, but zero tandem gradient signal early on.

**Proposed:** Coordinate them in opposite directions. Start with HIGH gap/stagger sigma (0.06) while tandem loss weight is LOW, then as tandem_ramp rises, simultaneously ramp sigma DOWN to 0.02. Intuition: broad geometric diversity early (helps learn general tandem priors), precise OOD-focused supervision late.

**Critical distinction from augmentation annealing dead end (#2152):** That experiment annealed augmentation PROBABILITY (turning augs off). This experiment changes sigma MAGNITUDE (maintaining constant probability but narrowing the perturbation range). Augmentation never turns off — it just focuses from broad to precise.

## Instructions

**Step 1 — Add CLI flags:**
```python
parser.add_argument('--coordinated_ramp', action='store_true',
                    help='Coordinate gap/stagger sigma with tandem_ramp')
parser.add_argument('--ramp_sigma_start', type=float, default=0.06,
                    help='Starting sigma (high diversity)')
parser.add_argument('--ramp_sigma_end', type=float, default=0.02,
                    help='Ending sigma (precise targeting)')
```

**Step 2 — In training loop, compute dynamic sigma each epoch.** Find where `aug_gap_stagger_sigma` is used to generate augmentation noise. Replace the fixed sigma with a dynamic one:

```python
if args.coordinated_ramp:
    # Get the tandem ramp fraction (same schedule that ramps tandem loss weight)
    # Look for where tandem_ramp_weight is computed — it goes from 0 to 1
    ramp_frac = tandem_ramp_weight  # 0 early, 1 late
    dynamic_sigma = args.ramp_sigma_start * (1 - ramp_frac) + args.ramp_sigma_end * ramp_frac
else:
    dynamic_sigma = args.aug_gap_stagger_sigma
```

Then use `dynamic_sigma` wherever `args.aug_gap_stagger_sigma` is used for noise generation.

**Important:** The tandem_ramp schedule is already in train.py — find how tandem_ramp_weight is computed per epoch and reuse the same fraction. The sigma should track it inversely: when tandem_ramp_weight is 0 (early), sigma is 0.06; when tandem_ramp_weight is 1 (late), sigma is 0.02.

**Step 3 — Run 2 configs, 2 seeds each** using `--wandb_group nezuko-coord-ramp`:

**Config A — sigma 0.06→0.02:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/coord-ramp-0.06" --wandb_group nezuko-coord-ramp \
  --coordinated_ramp --ramp_sigma_start 0.06 --ramp_sigma_end 0.02 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

**Config B — sigma 0.04→0.02 (narrower range):**
Same command with `--ramp_sigma_start 0.04` and `--wandb_name "nezuko/coord-ramp-0.04"`. (Repeat with `--seed 73`.)

Run all 4 in parallel. Report p_in, p_oodc, p_tan, p_re per config (2-seed avg).

**Note:** Log the dynamic_sigma value to W&B every epoch so we can see the decay curve.

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42), j9btfx09 (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Coordinated Tandem Ramp

### Implementation
- Added `--coordinated_ramp`, `--ramp_sigma_start`, `--ramp_sigma_end` flags
- Dynamic sigma computed per epoch: `sigma = start * (1 - ramp_frac) + end * ramp_frac`
- Ramp fraction reuses tandem_ramp schedule: `min(1.0, max(0.0, (epoch - 10) / 40.0))`
- Sigma logged to W&B as `train/dynamic_sigma` every step

### Sigma Schedule
| Epoch | ramp_frac | sigma (0.06→0.02) | sigma (0.04→0.02) |
|-------|-----------|--------------------|--------------------|
| 0 | 0.0 | 0.0600 | 0.0400 |
| 10 | 0.0 | 0.0600 | 0.0400 |
| 30 | 0.5 | 0.0400 | 0.0300 |
| 50 | 1.0 | 0.0200 | 0.0200 |
| 100+ | 1.0 | 0.0200 | 0.0200 |

### Results (Surface MAE — Pressure)

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B ID |
|--------|------|------|--------|-------|------|--------|
| σ 0.06→0.02 | 42 | 13.113 | 7.796 | 29.214 | 6.530 | `x2mkegnk` |
| σ 0.06→0.02 | 73 | 14.047 | 7.918 | 29.771 | 6.496 | `pvjzi2ov` |
| **σ 0.06→0.02 avg** | — | **13.580** | **7.857** | **29.493** | **6.513** | |
| σ 0.04→0.02 | 42 | 12.809 | 7.825 | 28.465 | 6.445 | `q5bprc0t` |
| σ 0.04→0.02 | 73 | 13.383 | 7.955 | 29.966 | 6.520 | `i4144o3d` |
| **σ 0.04→0.02 avg** | — | **13.096** | **7.890** | **29.216** | **6.483** | |
| **Baseline** | — | **13.05** | **7.70** | **28.60** | **6.55** | d7l91p0x, j9btfx09 |

### Analysis

**Neither config beats baseline on p_tan in the 2-seed average.** However, the results are nuanced:

**Config B (0.04→0.02) is much better than Config A (0.06→0.02).** The 3x starting sigma (0.06) is too aggressive — it hurts p_in by +4.1% and p_tan by +3.1%. The tighter 2x range (0.04→0.02) is nearly baseline-competitive.

**Standout run: coord-ramp-0.04-s42** individually beats baseline on 3 of 4 metrics:
- p_in: 12.81 vs 13.05 (-1.8%)
- p_tan: 28.47 vs 28.60 (-0.5%)
- p_re: 6.45 vs 6.55 (-1.6%)
But seed 73 regresses, so the 2-seed average doesn't beat baseline.

**High seed variance in p_tan (28.47 vs 29.97)** suggests the coordinated ramp adds instability to the tandem learning dynamics. The sigma transition happens during the critical tandem_ramp phase (epochs 10-50), and the concurrent changes may create a moving target that's hard to optimize consistently.

**p_re improves consistently** across both configs and seeds (6.48-6.51 vs 6.55), suggesting the early high-sigma phase helps Reynolds number generalization.

### Suggested Follow-ups
1. **Start the sigma ramp later (e.g., epoch 50-100)** to avoid interference with the tandem_ramp warmup phase
2. **Tighter range (0.03→0.02)** — the 0.04 starting sigma may still be too high
3. **Try coordinated ramp on dsdf2_sigma instead** — the geometry noise may benefit more from this schedule than gap/stagger noise

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2177 because ADVISOR: Closing — coordinated ramp doesn't reliably improve p_tan.

**Config A (0.06→0.02) avg:** p_tan=29.49 (+3.1%), p_in=13.58 (+4.1%)
**Config B (0.04→0.02) avg:** p_tan=29.22 (+2.2%), p_in=13.10 (+0.4%)

Config B seed 42 individually showed p_tan=28.47 (-0.5%), but seed 73 regressed to 29.97 (+4.8%). The 4.8% seed spread (vs 2.1% baseline) shows the concurrent sigma and loss ramp transitions create unstable optimization dynamics during the critical tandem warmup phase.

Your observation about p_re improving consistently is interesting — higher sigma early does help Re generalization. But the interference with tandem transfer learning makes the approach net negative.

The key takeaway: the tandem_ramp phase (epochs 10-50) is sensitive to other concurrent schedule changes. Any future schedule-based experiments should avoid modifying parameters during this critical window..
